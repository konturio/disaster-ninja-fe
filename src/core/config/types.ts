import type { UserDto } from '~core/app/user';
import type { AppFeatureType } from '~core/app/types';
import type { LayerDetailsDto } from '~core/logical_layers/types/source';
import type { UrlData } from '~core/url_store';

export type Config = {
  baseUrl: string;
  initialUrl: string;
  initialUrlData: UrlData;
  initialUser: UserDto;
  defaultLayers: LayerDetailsDto[];
  activeLayers: string[]; // Computed in boot stage
} & StageConfig &
  AppConfig;

export interface AppConfig {
  id: string;
  name: string;
  description: string;
  ownedByUser: boolean;
  features: FeaturesConfig;
  sidebarIconUrl: string;
  faviconUrl: string;
  faviconPack: FaviconPack;
  public: boolean;
  extent: [number, number, number, number];
  user?: UserDto;
}

export interface StageConfig {
  // Api gates
  apiGateway: string;
  reportsApiGateway: string;
  // Bivariate tiles
  bivariateTilesRelativeUrl: string;
  bivariateTilesIndicatorsClass: string;
  bivariateTilesServer?: string;
  // AutoRefreshService
  refreshIntervalSec: number;
  // Auth
  keycloakUrl: string;
  keycloakRealm: string;
  keycloakClientId: string;
  // Third-party services
  yandexMetricaId?: number[];
  intercomDefaultName?: string;
  intercomAppId?: string;
  intercomSelector?: string;
  // Event api config
  defaultFeed: string;
  // Map editor config
  osmEditors: OsmEditorConfig[];
  // Map config
  autofocusZoom: number;
  mapBlankSpaceId: string;
  mapBaseStyle: string;
  // App Defaults
  featuresByDefault: FeaturesConfig;
  defaultLanguage: string;
  // Sentry
  sentryDsn: string;
}

export interface OsmEditorConfig {
  id: string;
  title: string;
  url: string;
}

export interface EventFeedConfig {
  feed: string;
  name: string;
  description: string;
  default: boolean;
}

export type FeaturesConfig = Record<AppFeatureType, object | boolean>;
export interface FeatureDto {
  name: string;
  description: string;
  type: string;
  configuration: object;
}

export type FaviconPack = Record<
  | 'favicon.ico'
  | 'favicon.svg'
  | 'apple-touch-icon.png'
  | 'icon-192x192.png'
  | 'icon-512x512.png',
  string
>;

export type AboutFeatureConfig = {
  tabId: string;
  assetUrl: string;
  subTabs: {
    tabId: string;
    assetUrl?: string;
  }[];
};

export type LoginFeatureConfig = {
  profile_form: {
    organization_section: boolean;
    phone: boolean;
    linkedin: boolean;
  };
};
