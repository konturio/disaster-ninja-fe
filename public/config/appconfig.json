{
  "API_GATEWAY": "https://test-apps-ninja02.konturlabs.com/active/api",
  "GRAPHQL_API": "https://test-apps02.konturlabs.com/insights-api/graphql",
  "BOUNDARIES_API": "https://test-api02.konturlabs.com",
  "REPORTS_API": "https://test-apps-ninja02.konturlabs.com/active/reports",
  "BIVARIATE_TILES_RELATIVE_URL": "api/tiles/bivariate/v1/",
  "BIVARIATE_TILES_INDICATORS_CLASS": "all",
  "REFRESH_INTERVAL_SEC": 300,
  "MAP_ACCESS_TOKEN": "",
  "MAP_BASE_STYLE": "https://test-api02.konturlabs.com/layers/tiles/basemap/style_ninja.json",
  "LAYERS_BY_DEFAULT": [
    "BIV__Kontur OpenStreetMap Quantity",
    "activeContributors",
    "eventShape",
    "focused-geometry"
  ],
  "KEYCLOAK_URL": "https://keycloak01.konturlabs.com",
  "KEYCLOAK_REALM": "dev",
  "KEYCLOAK_CLIENT_ID": "kontur_platform",
  "AUTOFOCUS_PADDINGS": [16, 300, 16, 336],
  "AUTOFOCUS_ZOOM": 13,
  "INTERCOM_APP_ID": "e59cl64z",
  "FEATURES_BY_DEFAULT": [
    "events_list",
    "current_event",
    "reports",
    "current_episode",
    "episode_list",
    "osm_edit_link",
    "side_bar",
    "analytics_panel",
    "map_layers_panel",
    "focused_geometry_layer",
    "map_ruler",
    "boundary_selector",
    "draw_tools",
    "geometry_uploader",
    "legend_panel",
    "url_store",
    "feature_settings",
    "layers_in_area",
    "toasts",
    "interactive_map",
    "feed_selector",
    "header",
    "geocoder",
    "communities",
    "tooltip"
  ],
  "DEFAULT_FEED": "kontur-public"
}
