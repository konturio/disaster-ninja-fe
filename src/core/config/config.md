       /config/appconfig.json ->

url -> /apps/configuration -> getDefaultLayers -> getBasemapFromDetails -> getLayersDetails -> updateAppConfigOverrides
-> setAppLanguage

```ts
// globalThis.konturAppConfig
fetch(`${import.meta.env?.BASE_URL}config/appconfig.json`);

// +

const appConfigResponse = await apiClient.get<AppDto>(
  '/apps/configuration',
  { appId: initialState.app },
  true,
);
```

```ts

```
