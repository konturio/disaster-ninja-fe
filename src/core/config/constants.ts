export const ATLAS_APP_ID = '9043acf9-2cf3-48ac-9656-a5d7c4b7593d';
export const OASIS_APP_ID = '0b5b4047-3d9b-4ec4-993f-acf9c7315536';
