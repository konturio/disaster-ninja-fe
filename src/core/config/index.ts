import type { UrlData } from '~core/url_store';
import type { AppConfig, Config, StageConfig } from './types';
import type { UserDto } from '~core/app/user';
import type { LayerDetailsDto } from '~core/logical_layers/types/source';

class ConfigRepository {
  #config: Config | undefined;

  set({
    baseUrl,
    initialUrl,
    initialUrlData,
    stageConfig,
    appConfig,
    baseMapUrl,
    initialUser,
    defaultLayers,
    activeLayers,
  }: {
    baseUrl: string;
    initialUrl: string;
    initialUrlData: UrlData;
    stageConfig: StageConfig;
    appConfig: AppConfig;
    baseMapUrl: string;
    initialUser: UserDto;
    defaultLayers: LayerDetailsDto[];
    activeLayers: string[];
  }) {
    this.#config = {
      baseUrl,
      initialUrl,
      initialUrlData,
      ...stageConfig,
      ...appConfig,
      mapBaseStyle: baseMapUrl,
      features:
        Object.keys(appConfig.features).length > 0
          ? appConfig.features
          : stageConfig.featuresByDefault,
      initialUser,
      defaultLayers,
      activeLayers,
    };
  }

  get() {
    return this.#config!;
  }

  /* -- Intercom staff -- */

  #readSessionIntercomSetting = (key: string) =>
    sessionStorage.getItem(`kontur.intercom.${key}`);
  getIntercomSettings() {
    return {
      intercomAppId: this.#config!.intercomAppId,
      intercomSelector: this.#config!.intercomSelector,
      name: this.#readSessionIntercomSetting('name'),
      email: this.#readSessionIntercomSetting('email'),
      phone: this.#readSessionIntercomSetting('phone'),
    };
  }

  #setIntercomSetting = (k: string, v?: string) => {
    const key = `kontur.intercom.${k}`;
    v ? sessionStorage.setItem(key, v) : sessionStorage.removeItem(key);
  };
  updateIntercomSettings(settings: Record<string, string>) {
    Object.entries(settings).forEach(([k, v]) => this.#setIntercomSetting(k, v));
  }
}
export const configRepo = new ConfigRepository();
