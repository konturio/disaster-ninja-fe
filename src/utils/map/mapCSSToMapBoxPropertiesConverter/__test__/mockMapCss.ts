export const mapCSS = {
  width: 1,
  'casing-width': 3,
  color: '#FEFEFE',
  'casing-color': 'black',
  opacity: 0.5,
  'casing-opacity': 1,
  dashes: '1,2 ,4',
  'casing-dashes': '1',
  linecap: 'round',
  'casing-linecap': 'square',
  linejoin: 'miter',
  'casing-linejoin': 'bevel',
  'fill-color': 'rgb(255, 255, 255)',
  'fill-opacity': 0.3,
  'icon-image': 'airport',
  'icon-width': 13,
  'icon-height': 12,
  'icon-opacity': 1,
  'font-family': 'Open Sans, Tahoma',
  'font-size': 16,
  'font-weight': 'bold',
  'text-color': 'gray',
  'text-opacity': 1,
  'text-position': 'center',
  'text-offset': 3,
  'max-width': 300,
  text: 'label',
  'text-halo-color': 'red',
  'text-halo-radius': 2,
  offset: 3,
  'casing-offset': 4,
};
