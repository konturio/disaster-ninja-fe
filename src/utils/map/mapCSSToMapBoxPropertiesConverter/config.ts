// FORM https://wiki.openstreetmap.org/wiki/MapCSS/0.2
// TO https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/

/*
 * "casing_line" not exist in mapBox type,
 * but it required for generating additional layer of type "line" (one-type -> one-layer)
 */
export const MAP_CSS_MAPBOX = {
    // Line
    width: {
        type: 'line',
        category: 'paint',
        name: 'line-width',
        valueConverter: 'toNumber',
    },
    'casing-width': {
        type: 'casing_line',
        category: 'paint',
        name: 'line-width',
        valueConverter: 'relativeCasingLineWith',
    },
    color: {
        type: 'line',
        category: 'paint',
        name: 'line-color',
        valueConverter: null,
    },
    'casing-color': {
        type: 'casing_line',
        category: 'paint',
        name: 'line-color',
        valueConverter: null,
    },
    opacity: {
        type: 'line',
        category: 'paint',
        name: 'line-opacity',
        valueConverter: 'toNumber',
    },
    'casing-opacity': {
        type: 'casing_line',
        category: 'paint',
        name: 'line-opacity',
        valueConverter: 'toNumber',
    },
    dashes: {
        type: 'line',
        category: 'paint',
        name: 'line-pattern',
        valueConverter: 'splitByComma',
    },
    'casing-dashes': {
        type: 'casing_line',
        category: 'paint',
        name: 'line-pattern',
        valueConverter: 'splitByComma',
    },
    linecap: {
        type: 'line',
        category: 'layout',
        name: 'line-cap',
        valueConverter: null, // "round" and "square" as is
    },
    'casing-linecap': {
        type: 'casing_line',
        category: 'layout',
        name: 'line-cap',
        valueConverter: null, // "round" and "square" as is
    },
    linejoin: {
        type: 'line',
        category: 'layout',
        name: 'line-join',
        valueConverter: null, // "bevel" and "miter" as is
    },
    'casing-linejoin': {
        type: 'casing_line',
        category: 'layout',
        name: 'line-join',
        valueConverter: null, // "bevel" and "miter" as is
    },
    // fill
    'fill-color': {
        type: 'fill',
        category: 'paint',
        name: 'fill-color',
        valueConverter: null,
    },
    'fill-opacity': {
        type: 'fill',
        category: 'paint',
        name: 'fill-opacity',
        valueConverter: 'toNumber',
    },
    'fill-opacity-complex': {
        type: 'fill',
        category: 'paint',
        name: 'fill-opacity',
        valueConverter: null,
    },

    // symbol
    'icon-image': {
        type: 'symbol',
        category: 'layout',
        name: 'icon-image',
        valueConverter: null,
    },
    'icon-width': {
        type: 'symbol',
        category: 'layout',
        name: 'icon-size',
        valueConverter: 'toNumber',
    },
    'icon-height': {
        type: 'symbol',
        category: 'layout',
        name: 'icon-size', // overwrite icon-width
        valueConverter: 'toNumber',
    },
    'icon-opacity': {
        type: 'symbol',
        category: 'paint',
        name: 'icon-opacity',
        valueConverter: 'toNumber',
    },
    'icon-text-fit': {
        type: 'symbol',
        category: 'layout',
        name: 'icon-text-fit',
        valueConverter: null,
    },
    'text-field': {
        type: 'symbol',
        category: 'layout',
        name: 'text-field',
        valueConverter: null,
    },
    'text-allow-overlap': {
        type: 'symbol',
        category: 'layout',
        name: 'text-allow-overlap',
        valueConverter: null,
    },
    'text-variable-anchor': {
        type: 'symbol',
        category: 'layout',
        name: 'text-variable-anchor',
        valueConverter: null,
    },
    'symbol-placement': {
        type: 'symbol',
        category: 'layout',
        name: 'symbol-placement',
        valueConverter: null,
    },

    // font
    'font-family': {
        type: 'symbol',
        category: 'layout',
        name: 'text-font',
        valueConverter: 'splitByComma',
    },
    'text-font': {
        type: 'symbol',
        category: 'layout',
        name: 'text-font',
        valueConverter: 'splitByComma',
    },
    'font-size': {
        type: 'symbol',
        category: 'layout',
        name: 'text-size',
        valueConverter: 'toNumber',
    },
    'font-weight': {
        type: 'symbol',
        category: 'paint',
        name: 'text-halo-width',
        valueConverter: null,
    },
    // "font-style": { // * Not supported
    //   "type": "symbol",
    // },
    // "font-variant": { // * Not supported
    //   "type": "symbol",
    // },
    // "text-decoration": { // * Not supported
    //   "type": "symbol",
    // },
    'text-transform': {
        type: 'symbol',
        category: 'layout',
        name: 'text-transform',
        valueConverter: null,
    },
    'text-color': {
        type: 'symbol',
        category: 'paint',
        name: 'text-color',
        valueConverter: null,
    },
    'text-opacity': {
        type: 'symbol',
        category: 'paint',
        name: 'text-opacity',
        valueConverter: 'toNumber',
    },
    'text-position': {
        type: 'symbol',
        category: 'layout',
        name: 'symbol-placement',
        valueConverter: 'convertPlacement',
    },
    'text-offset': {
        type: 'symbol',
        category: 'layout',
        name: 'text-offset',
        valueConverter: 'convertOffset',
    },
    'max-width': {
        type: 'symbol',
        category: 'layout',
        name: 'text-max-width',
        valueConverter: 'toNumber',
    },
    text: {
        type: 'symbol',
        category: 'layout',
        name: 'text-field',
        valueConverter: null,
    },
    'text-halo-color': {
        type: 'symbol',
        category: 'paint',
        name: 'text-halo-color',
        valueConverter: null,
    },
    'text-halo-radius': {
        type: 'symbol',
        category: 'paint',
        name: 'text-halo-width',
        valueConverter: null,
    },
    'text-size': {
        type: 'symbol',
        category: 'layout',
        name: 'text-size',
        valueConverter: null,
    },
    // Extra
    offset: [
        {
            type: 'line',
            category: 'paint',
            name: 'line-offset',
            valueConverter: 'toNumber',
        },
        {
            type: 'casing_line',
            category: 'paint',
            name: 'line-offset',
            valueConverter: 'applyCasingOffset',
        },
    ],
    'casing-offset': null, // used only in value converter acceptCasingOffset
    'circle-color': {
        type: 'circle',
        category: 'paint',
        name: 'circle-color',
    },
    'circle-blur': {
        type: 'circle',
        category: 'paint',
        name: 'circle-blur',
    },
    'circle-opacity': {
        type: 'circle',
        category: 'paint',
        name: 'circle-opacity',
    },
    'circle-radius': {
        type: 'circle',
        category: 'paint',
        name: 'circle-radius',
    },
    'circle-stroke-color': {
        type: 'circle',
        category: 'paint',
        name: 'circle-stroke-color',
    },
    'circle-stroke-opacity': {
        type: 'circle',
        category: 'paint',
        name: 'circle-stroke-opacity',
    },
    'circle-stroke-width': {
        type: 'circle',
        category: 'paint',
        name: 'circle-stroke-width',
    },
    'circle-translate': {
        type: 'circle',
        category: 'paint',
        name: 'circle-translate',
    },
    'circle-translate-anchor': {
        type: 'circle',
        category: 'paint',
        name: 'circle-translate-anchor',
    },
};
