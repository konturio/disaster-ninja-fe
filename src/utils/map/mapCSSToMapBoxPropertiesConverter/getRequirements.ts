export function getRequirements(config, mapCSS) {
  const initialCSS = {
    // * Next important string need for generate casing offset when main line offset not set
    offset: 0,
  };

  const requirements: [unknown, unknown][] = [];
  Object.entries({ ...initialCSS, ...mapCSS }).forEach(([prop, value]) => {
    const req = config[prop];
    if (req === undefined) {
      console.debug(`Unsupported property: "${prop}"`);
    } else {
      requirements.push([req, value]);
    }
  });
  return requirements;
}
