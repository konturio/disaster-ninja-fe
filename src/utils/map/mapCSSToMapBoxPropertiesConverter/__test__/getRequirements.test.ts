import { test, expect } from 'vitest';
import { getRequirements } from '../getRequirements';
import { MAP_CSS_MAPBOX } from '../config';
import { mapCSS } from './mockMapCss';

test('Generate requirements from mapCSS', () => {
  const requirements = getRequirements(MAP_CSS_MAPBOX, mapCSS);
  expect(requirements).toMatchSnapshot();
});
