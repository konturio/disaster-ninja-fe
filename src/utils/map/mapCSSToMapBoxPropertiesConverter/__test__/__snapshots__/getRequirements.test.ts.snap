// Vitest Snapshot v1

exports[`Generate requirements from mapCSS 1`] = `
[
  [
    [
      {
        "category": "paint",
        "name": "line-offset",
        "type": "line",
        "valueConverter": "toNumber",
      },
      {
        "category": "paint",
        "name": "line-offset",
        "type": "casing_line",
        "valueConverter": "applyCasingOffset",
      },
    ],
    3,
  ],
  [
    {
      "category": "paint",
      "name": "line-width",
      "type": "line",
      "valueConverter": "toNumber",
    },
    1,
  ],
  [
    {
      "category": "paint",
      "name": "line-width",
      "type": "casing_line",
      "valueConverter": "relativeCasingLineWith",
    },
    3,
  ],
  [
    {
      "category": "paint",
      "name": "line-color",
      "type": "line",
      "valueConverter": null,
    },
    "#FEFEFE",
  ],
  [
    {
      "category": "paint",
      "name": "line-color",
      "type": "casing_line",
      "valueConverter": null,
    },
    "black",
  ],
  [
    {
      "category": "paint",
      "name": "line-opacity",
      "type": "line",
      "valueConverter": "toNumber",
    },
    0.5,
  ],
  [
    {
      "category": "paint",
      "name": "line-opacity",
      "type": "casing_line",
      "valueConverter": "toNumber",
    },
    1,
  ],
  [
    {
      "category": "paint",
      "name": "line-pattern",
      "type": "line",
      "valueConverter": "splitByComma",
    },
    "1,2 ,4",
  ],
  [
    {
      "category": "paint",
      "name": "line-pattern",
      "type": "casing_line",
      "valueConverter": "splitByComma",
    },
    "1",
  ],
  [
    {
      "category": "layout",
      "name": "line-cap",
      "type": "line",
      "valueConverter": null,
    },
    "round",
  ],
  [
    {
      "category": "layout",
      "name": "line-cap",
      "type": "casing_line",
      "valueConverter": null,
    },
    "square",
  ],
  [
    {
      "category": "layout",
      "name": "line-join",
      "type": "line",
      "valueConverter": null,
    },
    "miter",
  ],
  [
    {
      "category": "layout",
      "name": "line-join",
      "type": "casing_line",
      "valueConverter": null,
    },
    "bevel",
  ],
  [
    {
      "category": "paint",
      "name": "fill-color",
      "type": "fill",
      "valueConverter": null,
    },
    "rgb(255, 255, 255)",
  ],
  [
    {
      "category": "paint",
      "name": "fill-opacity",
      "type": "fill",
      "valueConverter": "toNumber",
    },
    0.3,
  ],
  [
    {
      "category": "layout",
      "name": "icon-image",
      "type": "symbol",
      "valueConverter": null,
    },
    "airport",
  ],
  [
    {
      "category": "layout",
      "name": "icon-size",
      "type": "symbol",
      "valueConverter": "toNumber",
    },
    13,
  ],
  [
    {
      "category": "layout",
      "name": "icon-size",
      "type": "symbol",
      "valueConverter": "toNumber",
    },
    12,
  ],
  [
    {
      "category": "paint",
      "name": "icon-opacity",
      "type": "symbol",
      "valueConverter": "toNumber",
    },
    1,
  ],
  [
    {
      "category": "layout",
      "name": "text-font",
      "type": "symbol",
      "valueConverter": "splitByComma",
    },
    "Open Sans, Tahoma",
  ],
  [
    {
      "category": "layout",
      "name": "text-size",
      "type": "symbol",
      "valueConverter": "toNumber",
    },
    16,
  ],
  [
    {
      "category": "paint",
      "name": "text-halo-width",
      "type": "symbol",
      "valueConverter": null,
    },
    "bold",
  ],
  [
    {
      "category": "paint",
      "name": "text-color",
      "type": "symbol",
      "valueConverter": null,
    },
    "gray",
  ],
  [
    {
      "category": "paint",
      "name": "text-opacity",
      "type": "symbol",
      "valueConverter": "toNumber",
    },
    1,
  ],
  [
    {
      "category": "layout",
      "name": "symbol-placement",
      "type": "symbol",
      "valueConverter": "convertPlacement",
    },
    "center",
  ],
  [
    {
      "category": "layout",
      "name": "text-offset",
      "type": "symbol",
      "valueConverter": "convertOffset",
    },
    3,
  ],
  [
    {
      "category": "layout",
      "name": "text-max-width",
      "type": "symbol",
      "valueConverter": "toNumber",
    },
    300,
  ],
  [
    {
      "category": "layout",
      "name": "text-field",
      "type": "symbol",
      "valueConverter": null,
    },
    "label",
  ],
  [
    {
      "category": "paint",
      "name": "text-halo-color",
      "type": "symbol",
      "valueConverter": null,
    },
    "red",
  ],
  [
    {
      "category": "paint",
      "name": "text-halo-width",
      "type": "symbol",
      "valueConverter": null,
    },
    2,
  ],
  [
    null,
    4,
  ],
  [
    {
      "category": "paint",
      "name": "circle-radius",
      "type": "circle",
    },
    3,
  ],
]
`;
